# Στοχαστικές Διεργασίες και Βελτιστοποίηση στη Μηχανική Μάθηση

## 3ο Εργαστήριο - *Εxercise 4*

- Ονομ/νυμο: Χρήστος Νίκου
- AM: 03400146
- Ιδιότητα: Μεταπτυχιακός φοιτητής Επιστήμης Δεδομένων και Μηχανικής Μάθησης (ΕΔΕΜΜ)
- Ηλεκτρονική Διεύθυνση: christosnikou@mail.ntua.gr / chrisnick92@gmail.com

<h1><b>Άσκηση 4</b></h1>
<p align="justify">Η μέθοδος Monte Carlo είναι μια υπολογιστική μέθοδος, που βασίζεται στο νόμο των μεγάλων αριθμών. Αν $(X_n)_{n\in \mathbb{N}}$ είναι μια ακολουθία από ανεξάρτητες, ισόνομες τυχαίες μεταβλητές, με πεπερασμένη μέση τιμή $Ε[Χ]$, τότε:</p>

$$
P\left[
\frac{1}{n}\sum_{k=1}^{n}X_k \rightarrow E[X]
\right] = 1
$$

<p align="justify">Προκειμένου να υπολογίσουμε τη μέση τιμή $Ε[Χ]$  μιας τυχαίας μεταβλητής $Χ$, μπορούμε λοιπόν να πάρουμε το μέσο όρο ενός μεγάλου αριθμού ανεξάρτητων δειγμάτων αυτής της μεταβλητής. Με παρόμοιο τρόπο, μπορούμε να προσεγγίσουμε υπολογιστικά την πιθανότητα ενός ενδεχομένου από το κλάσμα των πραγματοποιήσεών του σε μια σειρά από <b>m</b> ανεξάρτητες προσομοιώσεις μέχρι το βήμα <b>n</b>, δηλαδή:</p>

$$
P\left[
\frac{1}{m} \sum_{k=1}^{m}H_k \rightarrow P[X_n | X_0]
\right] = 1
$$

<p align="justify">όπου η τυχαία μεταβλητή $Η_k$ παίρνει την τιμή 1 εάν το ενδεχόμενο πραγματοποιείται στο τέλος του εκάστοτε πειράματος και 0 στην αντίθετη περίπτωση. Σ’ αυτήν την ιδέα θα βασιστεί η άσκηση αυτή. Σας δίνεται η μαρκοβιανή αλυσίδα στο χώρο καταστάσεων $Χ=\{1,2,3\}$ με πίνακα πιθανοτήτων μετάβασης:</p>

$$
P = \begin{pmatrix}
0 & 1 & 0\\
0 & 2/3 & 1/3\\
1/6 & 5/6 & 0
\end{pmatrix}
$$

<p align="justify">Χρησιμοποιώντας το πρόγραμμα που δίνεται παρακάτω, θα πραγματοποιήσετε <b>m</b> ανεξάρτητα πειράματα για να εκτιμήσετε την πιθανότητα <b>(Για να τρέξετε το πρόγραμμα θα πρέπει να έχετε φορτώσει το αρχείο <i><a href="https://github.com/nkostopoulos/StochasticsLabPublic/blob/master/lab3/simple_markov_chain_lib.py">simple_markov_chain_lib.py</a></i>)</b>.</p>

$$
P\left[
X_{40} = 1 | X_0 = 1
\right],
$$

<p align="justify">δηλαδή την πιθανότητα να βρίσκεται η αλυσίδα στην κατάσταση 1 στο 40ό βήμα της δεδομένου ότι ξεκίνησε από την κατάσταση 1. Για να ελέγξετε την ορθότητα της μεθόδου, το πρόγραμμα περιλαμβάνει και τον ακριβή υπολογισμό της παραπάνω πιθανότητας.</p> 
<ul>
<li>Να μελετήσετε το πρόγραμμα και να περιγράψετε σύντομα τη μέθοδο που ακολουθείται.</li>
<li>Να επαναλάβετε τη διαδικασία για τιμές της παραμέτρου m: (α) 1,000, (β) 10,000, (γ) 50,000, (δ) 100,000, (ε) 500,000. Να καταγράψετε και να σχολιάσετε την τιμή της παραπάνω πιθανότητας όπως υπολογίζεται από την προσομοίωση σε σχέση με την ακριβή τιμή της.</li>
</ul>


><b>1.</b>*Να μελετήσετε το πρόγραμμα και να περιγράψετε σύντομα τη μέθοδο που ακολουθείται.*

Η εκτίμηση της πιθανότητας $P\bigl(X_{40}=1 | X_{0}=1\bigr)$ μέσω της διαδικασίας *Monte Carlo* στο πρόγραμμα που ακολουθεί παρακάτω γίνεται μέσω της συνάρτησης *calculateProbabilities*. Η συνάρτηση αρχικοποιεί την αλυσίδα μέσω της μεταβλητής `mc` με πίνακα μεταβάσεων 
$$P  = \begin{pmatrix}
0 & 1 &  0 \\
0 & 2/3 & 1/3\\
1/6 & 5/6 & 0 \\
\end{pmatrix}.$$
Η μεταβλητή `expirements` δηλώνει τον αριθμό που θα επαναληφθεί το πείραμα, η μεταβλητή `steps` αρχικοποείται με την τιμή $40$ και αντιστοιχεί στον δείκτη $X_{40}$, δηλαδή ότι μας ενδιαφέρει η κατάσταση στην οποία βρίσκεται η αλυσίδα στο 40ο βήμα και τέλος η μεταβλητή `visits` είναι ο αριθμός των φορών που αλυσίδα βρέθηκε στην κατάσταση 1 κατά το 40ο βήμα. Μετά το τέλος των επαναλήψεων η εκτίμηση της ζητούμενης πιθανότητας γίνεται μέσω της 
$$P\bigl(X_{40}=1 | X_{0}=1\bigr)\approx \frac{visits}{experiments}.$$

Επιπλέον, στο παρακάτω πρόγραμμα υπολογίζεται και η πραγματική τιμή της πιθανότητας $P\bigl(X_{40}=1 | X_{0}=1\bigr)$ μέσω του τύπου $\pi^{(0)}\cdot P^{40}$, ο οποίος μας δίνει το διάνυσμα πιθανοτήτων της κάθε κατάστασης στο 40ο βήμα, έτσι η πρώτη συντεταγμένη αυτού του διανύσματος μας δίνει την πραγματική τιμή της πιθανότητας $P\bigl(X_{40}=1 | X_{0}=1\bigr)$.

In [2]:
from __future__ import division
from simple_markov_chain_lib import markov_chain
import statistics as stat
from numpy import matmul
import numpy as np

def defineMarkovTable(): 
	p = 1/6
	markov_table = {
		1: {2: 1.},
		2: {2: 2/3, 3: 1/3},
		3: {1: p, 2: 1-p}
	}

	return markov_table

def defineNumpyTable():
	Pn = np.array([[0,1,0],
		       [0,2/3,1/3],
		       [1/6, 5/6, 0]])
	P0 = np.array([[1,0,0]])
	
	return Pn,P0

def multiplyNumpyTables(Pn,P0):
	for index in range(40):
		Pn = np.matmul(Pn,Pn)
	Pn = np.matmul(P0,Pn)
	return Pn

def defineInitDistribution():
	init_dist = {1: 1.}
	
	return init_dist

def calculateProbabilities(markov_table, init_dist, experiments):
	mc = markov_chain(markov_table, init_dist)
	steps = 40
	visits = 0

	for index in range(experiments):
		mc.start()
		for j in range(steps):
			mc.move()
		if mc.running_state == 1: visits += 1

	probability = visits / experiments
	return probability, experiments

><b>2. </b>*Να επαναλάβετε τη διαδικασία για τιμές της παραμέτρου m: (α) 1,000, (β) 10,000, (γ) 50,000, (δ) 100,000, (ε) 500,000. Να καταγράψετε και να σχολιάσετε την τιμή της παραπάνω πιθανότητας όπως υπολογίζεται από την προσομοίωση σε σχέση με την ακριβή τιμή της.*

Παρακάτω εκτελούμε τη διαδικασία της μεθόδου *Monte Carlo* για (α) 1,000, (β) 10,000, (γ) 50,000, (δ) 100,000 και (ε) 500,000 επαναλήψεις και καταγράφουμε τα αποτελέσματα συγκρινοντάς τας με την πραγματική τιμή της πιθανότητας $P\bigl(X_{40}=1 | X_{0}=1\bigr)$.

#### *2. A - 1,000 Επαναλήψεις*

In [7]:
reps = 1000
if __name__ == "__main__":
	markov_table = defineMarkovTable()
	init_dist = defineInitDistribution()

	probability, experiments = calculateProbabilities(markov_table, init_dist, reps)
	print(f"- Estimated probability: {probability} using Monte Carlo method with total {experiments} experiments.")

	Pn,P0 = defineNumpyTable()
	realProbability = multiplyNumpyTables(Pn,P0)
	print(f"- Real probability: {realProbability[0][0]:.6f} using matrix multiplication")  
	print(f"- Probability vector in step 40: {realProbability}")

- Estimated probability: 0.035 using Monte Carlo method with total 1000 experiments.
- Real probability: 0.039999 using matrix multiplication
- Probability vector in step 40: [[0.03999857 0.71997421 0.2399914 ]]


#### *2. Β - 10,000 Επαναλήψεις*

In [8]:
reps = 10000
if __name__ == "__main__":
	markov_table = defineMarkovTable()
	init_dist = defineInitDistribution()

	probability, experiments = calculateProbabilities(markov_table, init_dist, reps)
	print(f"- Estimated probability: {probability} using Monte Carlo method with total {experiments} experiments.")

	Pn,P0 = defineNumpyTable()
	realProbability = multiplyNumpyTables(Pn,P0)
	print(f"- Real probability: {realProbability[0][0]:.6f} using matrix multiplication")  
	print(f"- Probability vector in step 40: {realProbability}")

- Estimated probability: 0.0372 using Monte Carlo method with total 10000 experiments.
- Real probability: 0.039999 using matrix multiplication
- Probability vector in step 40: [[0.03999857 0.71997421 0.2399914 ]]


#### *3. Γ - 50,000 Επαναλήψεις*

In [9]:
reps = 50000
if __name__ == "__main__":
	markov_table = defineMarkovTable()
	init_dist = defineInitDistribution()

	probability, experiments = calculateProbabilities(markov_table, init_dist, reps)
	print(f"- Estimated probability: {probability} using Monte Carlo method with total {experiments} experiments.")

	Pn,P0 = defineNumpyTable()
	realProbability = multiplyNumpyTables(Pn,P0)
	print(f"- Real probability: {realProbability[0][0]:.6f} using matrix multiplication")  
	print(f"- Probability vector in step 40: {realProbability}")

- Estimated probability: 0.04046 using Monte Carlo method with total 50000 experiments.
- Real probability: 0.039999 using matrix multiplication
- Probability vector in step 40: [[0.03999857 0.71997421 0.2399914 ]]


#### *4. Δ - 100,000 Επαναλήψεις*

In [10]:
reps = 100000
if __name__ == "__main__":
	markov_table = defineMarkovTable()
	init_dist = defineInitDistribution()

	probability, experiments = calculateProbabilities(markov_table, init_dist, reps)
	print(f"- Estimated probability: {probability} using Monte Carlo method with total {experiments} experiments.")

	Pn,P0 = defineNumpyTable()
	realProbability = multiplyNumpyTables(Pn,P0)
	print(f"- Real probability: {realProbability[0][0]:.6f} using matrix multiplication")  
	print(f"- Probability vector in step 40: {realProbability}")

- Estimated probability: 0.04183 using Monte Carlo method with total 100000 experiments.
- Real probability: 0.039999 using matrix multiplication
- Probability vector in step 40: [[0.03999857 0.71997421 0.2399914 ]]


#### *4. Ε - 500,000 Επαναλήψεις*

In [11]:
reps = 500000
if __name__ == "__main__":
	markov_table = defineMarkovTable()
	init_dist = defineInitDistribution()

	probability, experiments = calculateProbabilities(markov_table, init_dist, reps)
	print(f"- Estimated probability: {probability} using Monte Carlo method with total {experiments} experiments.")

	Pn,P0 = defineNumpyTable()
	realProbability = multiplyNumpyTables(Pn,P0)
	print(f"- Real probability: {realProbability[0][0]:.6f} using matrix multiplication")  
	print(f"- Probability vector in step 40: {realProbability}")

- Estimated probability: 0.039758 using Monte Carlo method with total 500000 experiments.
- Real probability: 0.039999 using matrix multiplication
- Probability vector in step 40: [[0.03999857 0.71997421 0.2399914 ]]


Όπως βλέπουμε όσο αυξάνεται ο αριθμός των επαναλήψεων τόσο η εκτιμώμενη πιθανότητα που παίρνουμε απ' τη μέθοδο *Monte Carlo* τείνει προς την πραγματική τιμή της πιθανότητας. Στον παρακάτω βλέπουμε και τα αποτελέσματα των εκτιμήσεων για την πιθανότητα $P\bigl(X_{40}=1 | X_{0}=1\bigr) = 0.039999.$

| **Αριθμός Επαναλήψεων** | **Εκτιμώμενη πιθανότητα**|
|:---------------|:-----------------|
| **1,000** |       0.035      |
| **10,000**      |       0.0372      |
| **50,000**   |      0.04046    |       
|  **100,000**	     |       0.04183   |
| **500,000**|       0.039758     |